climada/util/calibrate/impact_func.py

```diff
@@ -0,0 +1,111 @@
+"""Module for calibrating impact functions"""
+
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from typing import Callable, Mapping, Optional, Tuple, Union, Any
+from numbers import Number
+
+import numpy as np
+import pandas as pd
+from scipy.optimize import (
+    Bounds,
+    LinearConstraint,
+    NonlinearConstraint,
+    OptimizeResult,
+    minimize,
+)
+from bayes_opt import BayesianOptimization
+
+from ....climada.hazard import Hazard
+from ....climada.entity import Exposures, ImpactFunc, ImpactFuncSet
+from ....climada.engine import Impact, ImpactCalc
+
+
+@dataclass
+class Input:
+    """Define the static input for a calibration task"""
+
+    hazard: Hazard
+    exposure: Exposures
+    data: pd.DataFrame
+    cost_func: Callable[[Impact, pd.DataFrame], float]
+    impact_func_gen: Callable[..., ImpactFuncSet]
+    bounds: Optional[Mapping[str, Union[Bounds, Tuple[Number, Number]]]] = None
+    constraints: Optional[
+        Mapping[str, Union[LinearConstraint, NonlinearConstraint, Mapping]]
+    ] = None
+
+    def __post_init__(self):
+        """Prepare input data"""
+        self.hazard = self.hazard.select(event_id=self.data.index)
+        self.exposure.assign_centroids(self.hazard)
+
+@dataclass
+class Output:
+    """Define the output of a calibration task"""
+
+    params: Mapping[str, Number]
+    target: Number
+    success: bool
+    result: Optional[OptimizeResult] = None
+
+
+@dataclass
+class Optimizer(ABC):
```

Comment:
As I first thought of this class, I did not expect it to be a `dataclass`. Is there a strong reason to make it a dataclass instead of a normal class? The purpose of this class is to compute values right? Not to store data?

Reply:
`dataclass` relieves us from writing trivial, yet error-prone `__init__` methods. This is particularly important because we want to derive from `Optimizer` and might want to add more attributes in the derived class. Compare:
```python
@dataclass
class Optimizer(ABC):
    input: Input

@dataclass
class Derived(Optimizer):
    other_input: OtherInput

class OptimizerAlt(ABC):
    def __init__(input):
        self.input = input

class DerivedAlt(OptimizerAlt):
    def __init__(input, other_input):
        self.other_input = other_input
        super().__init__(input)

# Called the same way
Derived(my_input, my_other_input)
DerivedAlt(my_input, my_other_input)
```

Reply:
I see. I am a bit confused still. This is just how class inheritance is made in Python. Do you suggest using a dataclass just to handle the `__init__`, even though it is not a dataclass?

Reply:
I can't quite follow. It is a dataclass when I decorate it as such. The term "dataclass" might be a bit misleading, but the [docs](https://docs.python.org/3/library/dataclasses.html) just state:

> [dataclasses](https://docs.python.org/3/library/dataclasses.html#module-dataclasses) — Data Classes
>
> This module provides a decorator and functions for automatically adding generated [special method](https://docs.python.org/3/glossary.html#term-special-method)s such as __init__() and __repr__() to user-defined classes.

This is my whole intention here.

Reply:
I mean there are certainly different opinions on how to best make classes, but dataclasses are meant to be used to create classes that mostly simply store data. The proposed class `Optimizer` however is exactly the opposite, a class that does mostly process data. What I understand from your argument above is that all classes should always be with the decorator dataclass. 

If you could clarify when you would NOT use a dataclass, that might help. 

https://stackoverflow.com/questions/47955263/what-are-data-classes-and-how-are-they-different-from-common-classes

Reply:
I see that this is the intended use but I still feel like you are focusing on the fact that the decorator is called "dataclass" and you percive the class as not being a data class. [attrs](https://www.attrs.org/en/stable/examples.html) does the exact same thing (and more) using the `@define` decorator. Would you also argue against that? I feel my use of `@dataclass` is justified here. I just want to remove boilerplate code. I would argue that the use case for `dataclass` is that a class stores objects that are passed through its constructor with little to no modifications (hence resulting in a lot of trivial boilerplate code). Using `dataclass` to simplify that is independent from any other functionality of the class.

I was trying things out and I find the resulting code clearer. But we can switch back to the usual `__init__` at any time.

Reply:
We decided that using `dataclass` is fine and justified, especially because it reduces the amount of boilerplace code.